File: 0015-3sum/0015-3sum.cpp
class Solution {
public:

vector<vector<int>> threeSum(vector<int> &nums)
{
    vector<vector<int>> ans;
    set<vector<int>> st;
    int n = nums.size();
    sort(nums.begin(), nums.end());
    for (int i = 0; i < n; i++)
    {
        int j = i + 1, k = n - 1;
        while (j < k)
        {
            int sum = nums[i] + nums[j] + nums[k];
            if (sum == 0)
            {
                vector<int> triplets = {nums[i], nums[j], nums[k]};
                sort(triplets.begin(), triplets.end());
                st.insert(triplets);
                j++;k--;
            }
            else if (sum < 0)
            {
                j++;
            }
            else if (sum > 0)
                k--;
        }
    }
    for (auto i : st)
    {
        ans.push_back(i);
    }
    return ans;
}
};

File: 0015-3sum/0015-3sum.java
class Solution {
 public List<List<Integer>> threeSum(int[] nums) {
        Set<List<Integer>> st = new HashSet<>();
        int n = nums.length;
        for (int i = 0; i < n - 1; i++) {
            int target = -nums[i];
            HashMap<Integer, Integer> mp = new HashMap<>();
            mp.put(nums[i + 1], i);
            for (int j = i + 2; j < n; j++) {
                int req = target - nums[j];
                if (mp.get(req) != null) {
                    List<Integer> ds = new ArrayList<>(List.of(nums[i], nums[j], req));
                    ds.sort(null);
                    st.add(ds);
                }
                mp.put(nums[j], j);
            }
        }
        return st.stream().toList();
    }
}